lib/handler/configurator/proxy.c

```diff
@@ -281,32 +403,77 @@ static int on_config_reverse_url(h2o_configurator_command_t *cmd, h2o_configurat
         inputs = node->data.sequence.elements;
         num_upstreams = node->data.sequence.size;
         break;
+    case YOML_TYPE_MAPPING:
+        for (i = 0; i < node->data.mapping.size; i++) {
+            yoml_t *key = node->data.mapping.elements[i].key;
+            yoml_t *value = node->data.mapping.elements[i].value;
+            if (key->type != YOML_TYPE_SCALAR) {
+                h2o_configurator_errprintf(cmd, key, "key must be a scalar");
+                return -1;
+            }
+            if (strcasecmp(key->data.scalar, "backends") == 0) {
+                switch (value->type) {
+                case YOML_TYPE_SCALAR:
+                    inputs = &value;
+                    num_upstreams = 1;
+                    break;
+                case YOML_TYPE_SEQUENCE:
+                    inputs = value->data.sequence.elements;
+                    num_upstreams = value->data.sequence.size;
+                    break;
+                default:
+                    h2o_configurator_errprintf(cmd, value, "value for backends must be either a scalar or a sequence");
+                    return -1;
+                }
+                continue;
+            }
+            if (strcasecmp(key->data.scalar, "balancer") == 0) {
+                balancer_conf = value;
+            }
+        }
+        break;
     default:
         h2o_fatal("unexpected node type");
         return -1;
     }
-    h2o_url_t *upstreams = alloca(sizeof(*upstreams) * num_upstreams);
-    for (i = 0; i != num_upstreams; ++i) {
-        if (inputs[i]->type != YOML_TYPE_SCALAR) {
-            h2o_configurator_errprintf(cmd, inputs[i], "argument to proxy.reverse.url must be a scalar or a sequence");
-            return -1;
-        }
-        if (h2o_url_parse(inputs[i]->data.scalar, SIZE_MAX, upstreams + i) != 0) {
-            h2o_configurator_errprintf(cmd, inputs[i], "failed to parse URL: %s\n", inputs[i]->data.scalar);
+
+    h2o_socketpool_target_t **targets = alloca(sizeof(*targets) * num_upstreams);
+
+    if (inputs == NULL) {
+        h2o_configurator_errprintf(cmd, node, "No backend is defined.");
```

Comment:
Could we move the error check to `case YOML_MAPPING` and also return `-1` if the check fails?

I think that checking for an error only when it might happen and bailing out earlier lead to better readability.